components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt

```diff
@@ -42,48 +44,80 @@ class ToolbarPresenter(
         renderer.start()
 
         scope = store.flowScoped { flow ->
-            flow.map { state -> state.findCustomTabOrSelectedTab(customTabId) }
-                .ifChanged()
+            flow.ifAnyChanged { arrayOf(it.findCustomTabOrSelectedTab(customTabId), it.extensions) }
                 .collect { state ->
-                    if (state == null) {
-                        clear()
-                    } else {
-                        render(state)
-                    }
+                    render(state)
                 }
         }
     }
 
     fun stop() {
         scope?.cancel()
-
         renderer.stop()
     }
 
     @VisibleForTesting(otherwise = PRIVATE)
-    internal fun render(tab: SessionState) {
-        renderer.post(tab.content.url)
+    internal fun render(state: BrowserState) {
+        val tab = state.findCustomTabOrSelectedTab(customTabId)
 
-        toolbar.setSearchTerms(tab.content.searchTerms)
-        toolbar.displayProgress(tab.content.progress)
+        if (tab != null) {
+            renderer.post(tab.content.url)
 
-        toolbar.siteSecure = if (tab.content.securityInfo.secure) {
-            Toolbar.SiteSecurity.SECURE
-        } else {
-            Toolbar.SiteSecurity.INSECURE
-        }
+            toolbar.setSearchTerms(tab.content.searchTerms)
+            toolbar.displayProgress(tab.content.progress)
+
+            toolbar.siteSecure = if (tab.content.securityInfo.secure) {
+                Toolbar.SiteSecurity.SECURE
+            } else {
+                Toolbar.SiteSecurity.INSECURE
+            }
+
+            toolbar.siteTrackingProtection = when {
+                tab.trackingProtection.ignoredOnTrackingProtection -> SiteTrackingProtection.OFF_FOR_A_SITE
+                tab.trackingProtection.enabled && tab.trackingProtection.blockedTrackers.isNotEmpty() ->
+                    SiteTrackingProtection.ON_TRACKERS_BLOCKED
 
-        toolbar.siteTrackingProtection = when {
-            tab.trackingProtection.ignoredOnTrackingProtection -> SiteTrackingProtection.OFF_FOR_A_SITE
-            tab.trackingProtection.enabled && tab.trackingProtection.blockedTrackers.isNotEmpty() ->
-                SiteTrackingProtection.ON_TRACKERS_BLOCKED
+                tab.trackingProtection.enabled -> SiteTrackingProtection.ON_NO_TRACKERS_BLOCKED
 
-            tab.trackingProtection.enabled -> SiteTrackingProtection.ON_NO_TRACKERS_BLOCKED
+                else -> SiteTrackingProtection.OFF_GLOBALLY
+            }
 
-            else -> SiteTrackingProtection.OFF_GLOBALLY
+            renderWebExtensionActions(state, tab)
+        } else {
+            clear()
+        }
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    internal fun renderWebExtensionActions(state: BrowserState, tab: SessionState) {
+        val extensionsMap = tab.extensionState.toMutableMap()
+        extensionsMap.putAll(state.extensions)
```

Comment:
Doesn't this do the inverse of what you want? Start with the session specific ones and then override with the global ones?

Reply:
Thanks for the catch I really meant the other way around. I updated the pr and I added a test to verified the expected behavior.